avocado/core/remoter.py

```diff
@@ -222,10 +245,12 @@ def run(self, command, ignore_status=False, quiet=True, timeout=60):
         :rtype: :class:`avocado.utils.process.CmdResult`.
         :raise fabric.exceptions.CommandTimeout: When timeout exhausted.
         """
-        return_dict = fabric.tasks.execute(run, command, ignore_status,
-                                           quiet, timeout,
-                                           hosts=[self.hostname])
-        return return_dict[self.hostname]
+
+        with shell_env(**self.env_vars):
```

Comment:
Or maybe only use the `shell_env` if `env_keep` is not `None`... again, up to you.